_posts/2019-05-31-ucx-dgx.md

```diff
@@ -0,0 +1,531 @@
+
+    spec = {
+        i: {
+            "cls": Nanny,
+            "options": {
+                "env": {
+                    "CUDA_VISIBLE_DEVICES": cuda_visible_devices(
+                        ii, CUDA_VISIBLE_DEVICES
+                    ),
+                    # 'UCX_NET_DEVICES': 'mlx5_%d:1' % (i // 2)
+                    "UCX_TLS": "rc,cuda_copy,cuda_ipc",
+                },
+                "interface": interface + str(i // 2),
+                "protocol": "ucx",
+                "ncores": threads_per_worker,
+                "data": dict,
+                "preload": ["dask_cuda.initialize_context"],
+                "dashboard_address": ":0",
+                "plugins": [CPUAffinity(affinity[i])],
+                "silence_logs": silence_logs,
+                "memory_limit": memory_limit,
+            },
+        }
+        for ii, i in enumerate(CUDA_VISIBLE_DEVICES)
+    }
+
+    scheduler = {
+        "cls": Scheduler,
+        "options": {
+            "interface": interface + str(CUDA_VISIBLE_DEVICES[0] // 2),
+            "protocol": "ucx",
+            "dashboard_address": dashboard_address,
+        },
+    }
+
+    return SpecCluster(
+        workers=spec, scheduler=scheduler, silence_logs=silence_logs, **kwargs
+    )
+```
+
+However, we never got the NVLink structure down.  The Dask scheduler currently
+still assumes uniform bandwidths between workers.  We've started to make small
+steps towards changing this, but we're not there yet (this will be useful as
+well for people that want to think about in-rack or cross-data-center
+deployments).
+
+As usual, in solving a highly specific problem, we were able to solve a number
+of lingering general features, which then made our specific problem easy to
+write down.
+
+
+Future Work
+-----------
+
+-   **Routing within complex networks**:
+    If you restrict yourself to four of the eight GPUs in a DGX, you can get 5-10 GB/s
+    between pairs of GPUs.  For some workloads this can be significant.  It
+    makes the system feel much more like a single unit than a bunch of isolated
+    machines.
+
+    However we still can't get great performance across the whole DGX because
+    there are many GPU-pairs that are not connected by NVLink, and so we get 10x
+    slower speeds.  These dominate communication costs if you naively try to use
+    the full DGX.
+
+    This might be solved either by:
+
+    1.  Teaching Dask to avoid these communications
+    2.  Teaching UCX to route communications like these through a chain of
+        multiple NVLink connections
+    3.  Avoiding complex networks altogether.  Apparently newer systems like
+        the DGX-2 use NVSwitch, which is more uniform.
+
+-   **CPU:** We can get 1-2 GB/s across infiniband, which isn't bad, but also
+    wasn't the full 5-8 GB/s that we was hoping for.  This deserves more serious
+    profiling to determine what is going wrong.  The current guess is that this
+    has to do with memory allocations.
```

Comment:
Do we have a sample that I can use to reproduce/benchmark this? I would like to check that, we could try something less burdensome, such as using Cython, like the example I wrote in [this gist](https://gist.github.com/pentschev/cf616610646c9de5ab08e4dc130bd8b3).

Reply:
The example above should work if you drop the `RandomState=cupy.RandomState` parameter to `da.RandomState(...)`

Reply:
Something the the following might also be helpful and a bit simpler.

```python
async with DGX(asynchronous=True, CUDA_VISIBLE_DEVICES=[0, 1]) as cluster:
    async with Client(cluster, asynchronous=True, direct_to_workers=True) as client:
        x = np.random.random(100000000)
        future = await client.scatter(x, hash=False)
        y = await future
```